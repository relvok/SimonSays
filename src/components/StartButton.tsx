import React, {useState} from 'react';
import {Text, View, TouchableOpacity} from 'react-native';
import {useDispatch} from 'react-redux';
import {AppDispatch} from '../store';
import {StartButtonProps} from './types';
import styles from './styles';

const StartButton = ({
  sounds,
  pads,
  simonSequence,
  setGameState,
  gameState,
  playSequence,
}: StartButtonProps) => {
  const [disabled, setDisabled] = useState(false);
  const dispatch = useDispatch<AppDispatch>();

  const onClickHandler = async () => {
    dispatch(setGameState({started: true, turn: 'simon', gameOver: false}));
    setDisabled(true);
    await playSequence(simonSequence, pads, sounds);
  };

  return (
    <View style={styles.startContainer}>
      <TouchableOpacity
        testID="start-button"
        activeOpacity={1}
        disabled={!gameState.started && !gameState.gameOver}
        style={styles.startButton}
        onPress={() => onClickHandler()}>
        <Text style={styles.startText}>START</Text>
      </TouchableOpacity>
    </View>
  );
};

export default StartButton;
