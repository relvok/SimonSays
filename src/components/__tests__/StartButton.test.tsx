import React from 'react';
import StartButton from 'components/StartButton';
import renderer from 'react-test-renderer';
import {fireEvent, act} from '@testing-library/react-native';

describe('StartButton component', () => {
  const setShowStart = jest.fn();
  const playSequence = jest.fn();
  //render StartButton component
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <StartButton
          sounds={[]}
          pads={[]}
          simonSequence={[]}
          setShowStart={setShowStart}
          playSequence={playSequence}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls setShowStart when button is pressed', () => {
    const component = renderer.create(
      <StartButton
        sounds={[]}
        pads={[]}
        simonSequence={[]}
        setShowStart={setShowStart}
        playSequence={playSequence}
      />,
    );
    const instance = component.root;
    const button = instance.findByProps({
      testID: 'start-button',
    });
    // button.props.onPress();
    act(() => {
      fireEvent.press(button);
    });
    expect(setShowStart).toHaveBeenCalled();
  });
});
