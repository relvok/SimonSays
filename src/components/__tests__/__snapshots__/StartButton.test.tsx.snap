// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartButton component renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "elevation": 5,
      "justifyContent": "center",
      "position": "absolute",
      "zIndex": 1,
    }
  }
>
  <View
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 60,
        "elevation": 5,
        "height": 70,
        "justifyContent": "center",
        "opacity": 1,
        "width": 70,
      }
    }
    testID="start-button"
  >
    <Text
      style={
        Object {
          "color": "grey",
          "fontFamily": "Roboto",
          "fontSize": 15,
          "fontWeight": "bold",
        }
      }
    >
      START
    </Text>
  </View>
</View>
`;
